x/concentrated-liquidity/swapstrategy/one_for_zero.go

```diff
@@ -87,6 +87,31 @@ func (s oneForZeroStrategy) ComputeSwapWithinBucketOutGivenIn(sqrtPriceCurrent,
 	// Note that spread reward is always charged on the amount in.
 	spreadRewardChargeTotal := computeSpreadRewardChargePerSwapStepOutGivenIn(hasReachedTarget, amountOneIn, amountOneInRemaining, s.spreadFactor)
 
+	fmt.Println("amountOneIn", amountOneIn)
+	fmt.Println("amountOneInRemaining", amountOneInRemaining)
+	fmt.Println("sqrtPriceCurrent", sqrtPriceCurrent)
+	fmt.Println("sqrtPriceNext", sqrtPriceNext)
+
+	// This covers an edge case where due to the lack of precision, the difference between the current sqrt price and the next sqrt price is so small that
+	// it ends up being rounded down to zero. This leads to an infinite loop in the swap algorithm. From knowing that this is a case where !hasReachedTarget,
+	//(that is the swap stops within a bucket), we charge the full amount remaining in to the user and infer the amount out from the sqrt price truncated
+	// in favor of the pool.
+	if !hasReachedTarget && sqrtPriceCurrent.Equal(sqrtPriceNext) && amountOneIn.IsZero() && !amountOneInRemaining.IsZero() {
+		amountOneIn = amountOneInRemaining
+		// Subtract 1 ULP so that the amount out is rounded in favor of the pool.
+		// TODO: figure out precision difference in the denominations of two tokens,
+		// leading to incorrect spot price in out core logic.
+
+		// sqrt price is amount 1 / amount 0
+		// How many tokens 1 do we need to get 1 token 0?
+		priceOneOverZero := sqrtPriceCurrent.MulRoundUp(sqrtPriceCurrent)
+
+		// how many tokens 0 do we get for 1 token 1?
+		priceZeroOverOne := sdk.OneDec().QuoTruncate(priceOneOverZero)
+
+		amountZeroOut = amountOneIn.Mul(priceZeroOverOne)
```

Comment:
On more thinking, I don't think the precision difference would be an issue here. Although the spot price, would be incorrect due to the precision difference it does not affect the correctness of this calculation.

Assume:
- quote asset (token 1) of uosmo with precision 6 and reserves `218023341414`
- base asset (token 0) of DAI with precision 18 and reserves `101170077995723619690981`
- `amountOneIn` of `1_000_000_000_000`

Therefore, the true price of the pool is:
```python
>>> (218023341414 / 10**6)  / (101170077995723619690981 / 10**18)
2.1550180224553714
```

However, in our core logic, it is represented as:

```python
218023341414 / 101170077995723619690981
2.1550180224553714e-12
```

Therefore, in our logic here, we would calculate:
```
1 / 2.1550180224553714e-12 = 464033242218.8405

100 * 464033242218.8405 = 46403324221884.055
```

When the true value is:
```
1 / 2.1550180224553714 = 0.46403324221884046

100 * 0.46403324221884046 =46.403324221884046
```

This means that we would remove `46403324221884` from reserves that is in fact: `46403324221884 / 10**12 = 46.403324221884046`